improver/wxcode/modal_code.py

```diff
@@ -0,0 +1,173 @@
+                values that need to be replaced with a more definitive period
+                code.
+            cube:
+                The original input data. Data relating to unset points will be
+                grouped and the mode recalculated."""
+
+        undecided_points = zip(*np.where(modal.data == UNSET_CODE_INDICATOR))
+
+        for y, x in undecided_points:
+            data = cube.data[:, y, x].copy()
+
+            for _, codes in GROUPED_CODES.items():
+                default_code = sorted([code for code in data if code in codes])
+                if default_code:
+                    for code in codes:
+                        data[data == code] = default_code[0]
+            mode_result, counts = stats.mode(CODE_MAX - data)
+            modal.data[y, x] = CODE_MAX - mode_result
+
+    @staticmethod
+    def mode_aggregator(data: ndarray, axis: int):
+        """An aggregator for use with iris to calculate the mode along the
+        specified axis. If the modal value selected comprises less than 10%
+        of data along the dimension being collapsed, the value is set to the
+        UNSET_CODE_INDICATOR to indicate that the uncertainty was too high to
+        return a mode.
+
+        Args:
+            data:
+                The data for which a mode is to be calculated.
+            axis:
+                The axis / dimension over which to calculate the mode.
+
+        Returns:
+            The data array collapsed over axis, containing the calculated modes.
+        """
+        # Iris aggregators support indexing from the end of the array.
+        if axis < 0:
+            axis += data.ndim
+        # Aggregation coordinate is moved to the -1 position in initialisation;
+        # move this back to the leading coordinate
+        data = np.moveaxis(data, [axis], [0])
+        minimum_significant_count = 0.1 * data.shape[0]
+        mode_result, counts = stats.mode(CODE_MAX - data, axis=0)
+        mode_result[counts < minimum_significant_count] = (
+            CODE_MAX - UNSET_CODE_INDICATOR
+        )
+        return CODE_MAX - np.squeeze(mode_result)
+
+    def process(self, cubes: Union[CubeList, List[Cube]]):
+        """Calculate the modal weather code, with handling for edge cases.
+
+        Args:
+            cubes:
+                A list of weather code cubes at different times. A modal
+                code will be calculated over the time coordinate to return
+                the most comon code, which is taken to be the best
+                representation of the whole period.
+
+        Returns:
+            A single weather code cube with time bounds that span those of
+            the input weather code cubes.
+        """
+        # Handle case in which a single time is provided.
+        if len(cubes) == 1:
+            return cubes[0]
+
+        cube = MergeCubes()(cubes)
+        self.unify_day_and_night(cube)
+
+        mode_aggregator = Aggregator("mode", self.mode_aggregator)
```

Comment:
This statement belongs in `self.__init__`

Reply:
Why might this go in init? It's just using a function defined within the class, I've just assigned it to a variable to make it a bit tidier, but it could equally be written as:

```
        result = cube.collapsed("time", Aggregator("mode", self.mode_aggregator))
```
I'm happy to move it, but I don't currently understand why.

Reply:
My thinking is that this statement does not depend on any of the inputs to the plugin method, but instead is constructing the aggregator, so belongs in the method that constructs the plugin, hence __init__. Consider running the plugin a thousand times with the same settings: the aggregator would be constructed a thousand times, when it only needed constructing once.
Practically, it doesn't make any difference, so if you'd rather leave it here, then I won't argue any more.

Reply:
Okay, now I understand. It's only invoked once, so it doesn't matter, but I will shift it so I can use new fancy python 3.7 dataclasses.

Reply:
Class variables set up using the dataclass decorator cannot refer to methods in the plugin itself. This is achieved, using a __post_init__ method, which in this case is no less code than just writing an init, so I've done that.